controllers/routing/routing_controller.go

```diff
@@ -84,19 +85,17 @@ func (r *PlatformRoutingController) Reconcile(ctx context.Context, req ctrl.Requ
 
 	var errs []error
 
-	if controllerutil.AddFinalizer(sourceRes, metadata.Finalizers.Routing) {
-		if errUpdate := r.Update(ctx, sourceRes); errUpdate != nil {
-			return ctrl.Result{}, fmt.Errorf("failed adding finalizer: %w", errUpdate)
-		}
+	addFinalizerErr := r.addResourceFinalizer(ctx, req, sourceRes)
+
+	if addFinalizerErr != nil {
+		return ctrl.Result{}, fmt.Errorf("failed adding finalizer with retry: %w", addFinalizerErr)
```

Comment:
We can simplify it to scoped `if err`

Two small nits on top:

- I think it's enough to call it `addFinalizer`
- I don't think we need to mention "retry" in error message. This is an implementation detail that can change, and it will be a wrapped error if this happens.
- We can get `NamespacedName` from passed target (`sourceRes`), no need to pass `req` around.
```suggestion
if errFinalizer := r.addFinalizer(ctx, sourceRes); errFinalizer != nil {
		return ctrl.Result{}, fmt.Errorf("failed adding finalizer: %w", errFinalizer)
```